src/opencage.rs

```diff
@@ -0,0 +1,414 @@
+//! The [OpenCage Geocoding](https://geocoder.opencagedata.com/) provider.
+//!
+//! Please see the [API documentation](https://geocoder.opencagedata.com/api) for details.
+//! Note that rate limits apply to the free tier; the remaining daily quota can be retrieved
+//! Using the [`remaining_calls()`](struct.Opencage.html#method.remaining_calls) method.
+//! ### A Note on Coordinate Order
+//! This provider's API documentation shows all coordinates in `[Latitude, Longitude]` order.
+//! However, `Geocoding` requires input `Point` coordinate order as `[Longitude, Latitude]`
+//! `(x, y)`, and returns coordinates with that order.
+//!
+//! ### Example
+//!
+//! ```
+//! use geocoding::{Opencage, Point, Reverse};
+//!
+//! let oc = Opencage::new("dcdbf0d783374909b3debee728c7cc10".to_string());
+//! let p = Point::new(2.12870, 41.40139);
+//! let res = oc.reverse(&p);
+//! println!("{:?}", res.unwrap());
+//! ```
+use std::sync::{Arc, Mutex};
+
+use super::num_traits::Float;
+use std::collections::HashMap;
+
+use super::Deserialize;
+use super::UA_STRING;
+use super::reqwest;
+use super::{header, Client};
+
+use super::Point;
+use super::{Forward, Reverse};
+
+// OpenCage has a custom rate-limit header, indicating remaining calls
+header! { (XRatelimitRemaining, "X-RateLimit-Remaining") => [i32] }
+
+/// An instance of the Opencage Geocoding service
+pub struct Opencage {
+    api_key: String,
+    client: Client,
+    endpoint: String,
+    remaining: Arc<Mutex<Option<i32>>>,
+}
+
+impl Opencage {
+    /// Create a new OpenCage geocoding instance
+    pub fn new(api_key: String) -> Self {
+        let mut headers = header::Headers::new();
+        headers.set(header::UserAgent::new(UA_STRING));
```

Comment:
So if a developer would want to change their `User-Agent` string they would need to make their opencage object mutable and then go through its client instance? Is there a better way about this?

Reply:
The OpenCage [developer guidelines](https://github.com/OpenCageData/opencagedata-roadmap/blob/master/library-guidelines.md) ask for a unique UA string, so they can track library usage, so this was a deliberate decision. I could add a getter / setter with the same interior mutability technique as `remaining` so that a `mut` `OpenCage` instance isn't necessary, but I'd also have to replace the `Client` instance in the same way, since its settings are retained across calls. The same applies to `endpoint` – its replacement or alteration requires a new `Client`. What do you think?

Reply:
Sounds fine for now, thanks for the explanation.